benchexec/tools/igml.py

```diff
@@ -0,0 +1,136 @@
+# https://github.com/sosy-lab/benchexec
+#
+# SPDX-FileCopyrightText: 2007-2020 Dirk Beyer <https://www.sosy-lab.org>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+import os
+import re
+import sys
+
+from numpy.matlib import identity
+
+import benchexec.result as result
+import benchexec.tools.template
+import benchexec.tools.template
+from benchexec.tools.sv_benchmarks_util import get_data_model_from_task, ILP32, LP64
+
+
+class Tool(benchexec.tools.template.BaseTool2):
+    """Tool info for IGML
+    The tool will be available soon"""
+
+    REQUIRED_PATHS = ["lib/*.jar", "scripts", "igml.jar"]
+
+    def executable(self, tool_locator):
+        executable = tool_locator.find_executable("igml.sh", subdir="scripts")
+        base_dir = os.path.join(os.path.dirname(executable), os.path.pardir)
+        jar_file = os.path.join(base_dir, "igml.jar")
+        bin_dir = os.path.join(base_dir, "target")
+        src_dir = os.path.join(base_dir, "src")
+        return executable
+
+    def program_files(self, executable):
+        return self._program_files_from_executable(
+            executable, self.REQUIRED_PATHS, parent_dir=True
+        )
+
+    def version(self, executable):
+        return self._version_from_tool(executable)
+
+    def name(self):
+        return "IGML"
+
+    def _get_additional_options(self, existing_options, task, rlimits):
+        options = []
+        if rlimits.cputime and "-timelimit" not in existing_options:
+            options += ["-timelimit", str(rlimits.cputime)]
+
+        if task.property_file:
+            options += ["-p", task.property_file]
+
+        # try:
+        #     options += ["-f", task.single_input_file]
+        # except benchexec.tools.template.UnsupportedFeatureException:
+        #     raise benchexec.tools.template.UnsupportedFeatureException(
+        #         "Unsupported task with multiple input files for task '{}'".format(task)
+        # )
+
+        if isinstance(task.options, dict) and task.options.get("language") == "C":
+            data_model = task.options.get("data_model")
+            if data_model:
+                data_model_option = get_data_model_from_task(
+                    task, {ILP32: "-m 32", LP64: "-m 64"}
+                )
+                if data_model_option:
+                    if data_model_option not in existing_options:
+                        options += [data_model_option]
+                else:
+                    raise benchexec.tools.template.UnsupportedFeatureException(
+                        "Unsupported data_model '{}' defined for task '{}'".format(
+                            data_model, task
+                        )
+                    )
+        return options
+
+    def cmdline(self, executable, options, task, rlimits):
+        additional_options = self._get_additional_options(options, task, rlimits)
+        return [executable] + options + additional_options
```

Comment:
This completely misses the actual input file(s) of the given task, is this intended?

Reply:
Yes, as the yml file is given to the tool, where the task is extracted from

Reply:
Where would the yml file be given to the tool? I don't see this either in the code. And in fact, BenchExec intentionally hides the name and full contents of the yml file from the tool-info module and the tool, because the yml file may contain the expected verdict.

Are you confusing this with the property file (which could in principle contain input files, but does not in environments like SV-COMP and Test-Comp, which given that you import `sv_benchmarks_util` seems likely to be your target use case)?

Reply:
I hand the filename of the yml file to the tool via the benchmark-definition (` <option name="-f">${taskdef_path_abs}/${taskdef_name}</option>`).

The tool contains two parts. A invariant generator and a validator.
Internally, the validator is given the .yml file, extracts the expected verdict and hand the task from the yml file to the invariant generator.
Thus, I want to give the expected verdict to the tool by purpose.

This may not be the best way to handle this and may abuse Benchexec.
(But as the tool is not designed to participate in the SV-COMP, this is in my point of view ok.

(A cleaner way would be to split the invariant generator and the verifier into two separate tools, but i am currently not able to do the refactoring needed. )

Reply:
I understand.